### PR TITLE
Compress conway polynomials

### PR DESCRIPTION
**Pull request by albinahlback**
- Conway polynomial database is now compressed by 89%.
- New function for getting Conway polynomials: `_nmod_poly_conway`
- New function to randomize degree and prime for Conway polynomials: `_nmod_poly_conway_rand` (the image of this function is isomorphic to the polynomials generated by `nmod_poly_conway`)
- ~Change `#ifdef FLINT_WANT_ASSERT` to `#if FLINT_WANT_ASSERT`.~ (moved to #1713)
- Optimize context initializers for `fq`, `fq_zech`, `fq_nmod` and `qadic` (also minimize/remove duplicate code between/in these modules).

Edit: I added the script I used to convert the old database to the new database. Basically, I had a hypothesis of how all polynomials should look within certain ranges, and then I checked that this held this during the conversion.

### CODE REVIEW

File: doc/source/nmod_poly.rst [hunk 1]
```diff
@@ -2294,6 +2294,26 @@ of polynomial multiplication.
...
+.. function:: int _nmod_poly_conway(mp_ptr op, ulong prime, slong deg)
+
+    Sets ``op`` to the coefficients to the Conway polynomial `C_{p, d}`, where
+    `p` is ``prime`` and `d` is ``deg``. This is done by checking against Frank
+    Lübeck's database [Lüb2004]_, which has been compressed in FLINT. Returns
+    `1` in case of success and returns `0` in case of failure.
+
+.. type:: struct _prime_degree_struct
+
+.. function:: struct _prime_degree_struct _nmod_poly_conway_rand(flint_rand_t state)
```
Comment (fredrik-johansson) [inline doc/source/nmod_poly.rst:2309]: I wouldn't introduce a struct for this; I would return one value and pass the other via a pointer.

**Comment by fredrik-johansson** [comment]
You could test that the entire database is correct (not just specific samples) by computing a hash over all the Conway polynomials.

**Comment by albinahlback** [comment]
> You could test that the entire database is correct (not just specific samples) by computing a hash over all the Conway polynomials.

I am 99% sure that the conversion is correct. But something is not right as the tests takes forever.

**Comment by albinahlback** [comment]
Ah, perhaps the reason is that now some `fq*_randtest` tests with bigger primes and bigger degrees. Let me check if this indeed is the case.

**Comment by albinahlback** [comment]
I think it is done. I optimized some test parameters as well in multiple modules.

I deprecated `fq*_ctx_randtest` and changed it to the more suitable `fq*_ctx_init_randtest`. It now accepts a final parameter `type` which determines what range the prime and degree will be in. This is documented.

**Comment by albinahlback** [comment]
Hopefully the tests won't take as long now.

**Comment by fredrik-johansson** [comment]
Did you time running tests before and after? Changed test parameters should be fine as long as test coverage hasn't gone down.

**Comment by albinahlback** [comment]
> Did you time running tests before and after?

No, but as I increased the range of random values for the finite fields, I needed to optimize some stuff to make it go fast again. Couldn't increase the random values as much as I'd like.

> Changed test parameters should be fine as long as test coverage hasn't gone down.

Not really, but I get what you mean. My thought process was the following:
- For alias checks, just check a couple of iterations. We do not need to check super duper many iterations to check that it doesn't do anything wrong.
- For functions relying on other functions (such as `fq_nmod` relying on `nmod_poly`), we can reduce the number of iterations and test parameters for a couple of values.
- Be careful around the basics (think of `mpn_mul_basecase`, where all possible carries are important to check).

**Comment by albinahlback** [comment]
But I get some weird error from testing `fq_nmod_poly_powmod_x_fmpz_preinv`. I did increase the number of iterations there by 10x, but I have no idea where the problem comes from. I did not change anything related to the function I believe, only the range of the random values...

**Comment by albinahlback** [comment]
Looks like the timings are similar to before, which is good as each iteration in the finite field section is expected to run a little bit longer now.

**Comment by albinahlback** [comment]
I think I see the problem. It is a bug in the main branch.

**Comment by albinahlback** [comment]
Here is a more detailed failure:
```
fq_nmod_poly_powmod_x_fmpz_preinv...
FAIL:
exp: 0
res1 is zero: 0
res2 is zero: 1
res1 is one: 1
res2 is one: 0
len(res1): 1
len(res2): 0
a:
2  0 211 1 211  1

f:
1  8 211  34 116 113 206 42 122 151 166

res1:
1  1 211  1

res2:
0
```

**Comment by fredrik-johansson** [comment]
The fq poly powmod functions aren't handling special cases correctly; they should handle exponent 0 (and maybe 1) before checking for cases like lenf == 1 and lenf == 2.

File: doc/source/nmod_poly.rst (hunk 1 above)
Comment (albinahlback) [reply doc/source/nmod_poly.rst]: This is how I prefer it. It is also a super duper very minor optimization (at least on x86 with System V ABI). Let me know if you still want me to change it.

Comment (fredrik-johansson) [reply doc/source/nmod_poly.rst]: The problem is just that a user who wants to call this function will have to figure out the struct members, and now there's suddenly a more things that ought to be documented than a single function. Besides, this isn't how most functions in FLINT handle returning a tuple (see ``fmpz_get_d_2exp``, ``n_gcdinv``, etc), so it goes a bit against convention. For a single use, a whole extra datatype doesn't seem worth it. No objections to using such private structs for static methods though.

Comment (albinahlback) [reply doc/source/nmod_poly.rst]: Will fix it!

**Comment by fredrik-johansson** [comment]
Positive review, once the test failure is fixed.

The issue might be that 1 modulo a constant polynomial should be 0.

**Comment by albinahlback** [comment]
It still randomly fails in Nemo.

**Comment by albinahlback** [comment]
I redid my `flint3` Nemo-branch. The previous test that failed on Nemo now passes on my computer (did some while loop, and when it had passed 30 seconds or so without failing I aborted it). Will now redo the Nemo test and then merge if it passes.

**Pull request merged**